### ISSUE TITLE
Json Error

### ISSUE THREAD

**Issue by rinaha** [2dvisio/cordova-plugin-googlefit#7]
Hi all,

We are trying to implemet GoogleFit in android and we get json error results.
We can see that the Google Fit sent us the request to add the google account and asking the the permission to get the data via apli. 

Even we give the  permission to the apli correctly, we have seeing the json error.

When we try to create the credientials, there are three options on the drop down.
1. API ley
2. OAuth client ID
3. Service account key

We have selected the OAuth client ID then we got the json error. We have tried with API just in case, but we got the same json error. 

We believe that we are doing ok by selecting "OAuth client ID", but we are not really sure.

If there are any advice regarding to this issue, please leave  us any comments.

Thank you!

**Comment by 2dvisio** [comment]
Hello @rinaha, have you tried the previous commit?
Maybe the problem was introduced with the last commit?

If you did it will help me debugging the problem :)

**Comment by rinaha** [comment]
Hi @2dvisio,

Thank you very much for your reply.
Are you refering "the previous commit" as our repository?
If  so, we do not have our own repository, unfortunately we are not doing the version control at this moment.

We are not sure if you can find anything from the codes below, but let us share it with you.

Please let us know if you need more information...

```
var gfit = {
    //alias plugin
    _plugin: null,
    //DataType to colletct
    _DataType: {
        "TYPE_STEP_COUNT_DELTA": "com.google.step_count.delta",
        "AGGREGATE_STEP_COUNT_DELTA": "com.google.step_count.delta",
        "TYPE_CALORIES_EXPENDED": "com.google.calories.expended",
        "AGGREGATE_CALORIES_EXPENDED": "com.google.calories.expended",
        "TYPE_DISTANCE_DELTA": "com.google.distance.delta",
        "AGGREGATE_DISTANCE_DELTA": "com.google.distance.delta",
        "TYPE_HEIGHT": "com.google.height",
        "TYPE_ACTIVITY_SAMPLE": "com.google.activity.sample",
        "TYPE_ACTIVITY_SEGMENT": "com.google.activity.segment",
        "AGGREGATE_ACTIVITY_SUMMARY": "com.google.activity.summary",
        "TYPE_WEIGHT": "com.google.weight",
        "TYPE_HEART_RATE_BPM": "com.google.heart_rate.bpm",
        "TYPE_POWER_SAMPLE": "com.google.power.sample",
        "AGGREGATE_POWER_SUMMARY": "com.google.power.summary",
        "TYPE_CYCLING_PEDALING_CADENCE": "com.google.cycling.pedaling.cadence",
        "TYPE_CYCLING_WHEEL_REVOLUTION": "com.google.cycling.wheel_revolution.rpm",
        "TYPE_LOCATION_SAMPLE": "com.google.location.sample"
    },
    //DurationUnit
    _DurationUnit: ["DAYS", "HOURS", "MINUTES", "SECONDS"],
    //SaveResults
    results: null,
    /**
     * CallBackGettingError
     **/
    _callback: function (res) { console.log(res); },
    initialize: function () {
        this._plugin = window.plugins.googlefit;
        this._callback = function (res) { console.log(res); };
    },
    /**
     * CallingGoogleFit
     * @param Object {
            start   : Integer   Start time in milliseconds
            end     : Integer   End time in milliseconds
            datatype: String    Datatypes under the URL format specified by GoogleFit
            callback: function  Call back after the value acquisition by GoogleFit
        }
     **/
    getData1: function (opts) {
        opts = this._normalize(opts);
        this._plugin.getStuff1(
                        opts.start,
                        opts.end,
                        this._DataType[opts.datatype],
                        this._success,
                        this._failed
                    );
    },
    /**
     * CallingGoogleFit
     * @param Object {
            start   : Integer   Start time in milliseconds
            end     : Integer   End time in milliseconds
            type1   : String    Datatypes under the URL format specified by GoogleFit
            type2   : String    Datatypes under the URL format specified by GoogleFit
            callback: function  Call back after the value acquisition by GoogleFit
            duration: Integer   Duration value of the databucket
            unit    : Integer   TimeUnit that quantify the duration unit (0: DAYS, 1: HOURS, 2: MINUTES, 3: SECONDS)_DurationUnit
            buckets : Integer   Type of the Buckets (0: ByTime, 1: ByActivityType, 2: ByActivitySegment)
        }
     **/
    getData2: function (opts) {
        opts = this._normalize(opts);
        this._plugin.getStuff2(
                        opts.start,
                        opts.end,
                        this._DataType[opts.type1],
                        this._DataType[opts.type2],
                        this.duration,
                        this._DurationUnit[opts.unit],
                        this.buckets,
                        this._success,
                        this._failed
                    );
    },
    /**
     * \Initialize the option value in order to get the value
     **/
    _normalize: function (opts) {
        if (isEmpty(this._plugin)) { this.initialize(); }
        if (isEmpty(opts)) { opts = {}; }
        if (isset(opts._callback)) { this._callback = opts.callback; }
        if (isEmpty(opts.datatype)) { opts.datatype = "TYPE_STEP_COUNT_DELTA"; }
        if (isEmpty(opts.type1)) { opts.type1 = "TYPE_STEP_COUNT_DELTA"; }
        if (isEmpty(opts.type2)) { opts.type2 = "TYPE_STEP_COUNT_DELTA"; }
        if (isEmpty(opts.duration)) { opts.duration = 1; }
        if (isEmpty(opts.unit)) { opts.unit = 0; }
        if (isEmpty(opts.buckets)) { opts.buckets = 0; }
        if (isEmpty(opts.start) || isEmpty(opts.end)) {
            var times = this._today();
            opts.start = times.start;
            opts.end = times.end;
        }
        return opts;
    },
    _today: function () {
        var now = new Date();
        var today = new Date(now.getFullYear() + "/" + (now.getMonth() + 1) + "/" + now.getDate());
        return { start: (+today), end: (+now) };
    },
    /**
     * CallBackFromGoogleFit(call async)
     **/
    _success: function (data) {
        gfit.results = data;
        console.log("success get fit data");

        gfit._callback(data);
    },
    /**
     *  CallBackFromGoogleFit(call async)
     **/
    _failed: function (e) {
        gfit.results = null;
        gfit._callback(e);
    }
};
```

**Comment by 2dvisio** [comment]
Hello @rinaha , no, I am referring to the previous commit of the this repository (the plugin).
The commit is this one: dc2df6889908d276ca56b777cbe348b3cbf5f695
It fixed issues with Marhsmellow new method for asking permissions, but might have caused issues somewhere else. 
Let me know if you manage to try the previous version and if you still have problem the cause is probably somewhere else. Otherwise, please do let me know and I will revise the code of the last commit to fix the bug.

Cheers.

**Comment by rinaha** [comment]
Thank you very much for your reply @2dvisio, 

We have tried the commit that you suggested us before, but unfortunately it made us to fail to build.  
This is what we get;

```
:compileReleaseJavaWithJavac/ProjectHomeDirectory/platforms/android/src/uk/ac/ox/ibme/cordova/plugins/GoogleFit.java:285: Error: Package "Manifest" does not exist
        if (ContextCompat.checkSelfPermission(actContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                                                                  ^
/ProjectHomeDirectory/platforms/android/src/uk/ac/ox/ibme/cordova/plugins/GoogleFit.java:285: Error: Not find symbol
        if (ContextCompat.checkSelfPermission(actContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ^
  symbol:   Variable ContextCompat
  position: Class GoogleFit
/ProjectHomeDirectory/platforms/android/src/uk/ac/ox/ibme/cordova/plugins/GoogleFit.java:285: Error: Not find symbol
        if (ContextCompat.checkSelfPermission(actContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                                                                                                       ^
  symbol:   Variable PackageManager
  position: Class GoogleFit
/ProjectHomeDirectory/platforms/android/src/uk/ac/ox/ibme/cordova/plugins/GoogleFit.java:288: Error: Package "Manifest" does not exist
                new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                                     ^
/ProjectHomeDirectory/platforms/android/src/uk/ac/ox/ibme/cordova/plugins/GoogleFit.java:287: Error: Not find symbol
        ActivityCompat.requestPermissions(actContext,
        ^
  symbol:   Variable ActivityCompat
  position: Class GoogleFit

5 of Error FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileReleaseJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 8.49 secs
Error: Error code 1 for command: /ProjectHomeDirectory/platforms/android/gradlew with args: cdvBuildRelease,-b,/ProjectHomeDirectory/platforms/android/build.gradle,-Dorg.gradle.daemon=true,-Pandroid.useDeprecatedNdk=true
```

We think that the problem comes from;

```
Error: Package "Manifest" does not exist here;
                new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
```

Also please note that we are using;
`<uses-sdk android:minSdkVersion="17" android:targetSdkVersion="22" />`

Thank you again for your prompt reply and looking for your response.
If we need to provide any additional information regarding our failed builds, please let us know.

Best regards,
Rina

**Comment by rinaha** [comment]
Hi @2dvisio,

We have been waiting for your reply. 

Please let us know if you need more information to solve this issue.

Thank you in advance.

Regards,
Rina

**Comment by 2dvisio** [comment]
Hello @rinaha that bit you are mentioning:

```
Error: Package "Manifest" does not exist here;
                new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
```

belongs to the latest commit.

Can you try with the commit previous to that? 

If that fixes it, it means that there is a problem with the last commit...

Thank you for trying.

**Comment by rinaha** [comment]
Hi @2dvisio,

Thank you for your reply.

Just to make sure,  the commit that we are supposed to use is this java file, https://github.com/2dvisio/cordova-plugin-googlefit/commit/dc2df6889908d276ca56b777cbe348b3cbf5f695 )
isn't it? 

Regarding this file contents,
`/android/GoogleFit.java` and `/src/android/GoogleFit.java` are very similar, but are these different ones?

We replaced `/android/GoogleFit.java` to `/src/android/GoogleFit.java`,
but would you recommend to make it the same hierarchical stracture as an original branch?

We have done the build successfully by making the same hierarchical stracture as an original branch, however, we are not be able to get the information from the plugin because of the JSONError.

Let us share the minimum project as a sample.
https://github.com/Sieg-sol-fried/cordova_gfit_sample

Also we have picked the codes only a part from the sample, have we done anything wrong here?

/www/js/index.js

```
    var start = (+new Date("2016/04/01")), end = (+new Date());
    window.plugins.googlefit.getStuff1(
                                        start,                          // Start time in milliseconds
                                        end,                            // Start time in milliseconds
                                        "com.google.step_count.delta",  // Datatypes under the URL format specified by GoogleFit
                                        function(data) {
                                            console.log(data);
                                        },
                                        function(e) {
                                            console.log(e);
                                        }
                                    );

```

One more issue that we would like to know, 
When we try to create the credientials, there are three options on the drop down.
1. API ley
2. OAuth client ID
3. Service account key

We have selected the OAuth client ID then we got the json error. We have tried with API just in case, but we got the same json error.

We are not sure which one is the correct way to create the credientials.

OAuth_client;
![oauth_client_gfit](https://cloud.githubusercontent.com/assets/16888650/16792035/73327be0-48ff-11e6-8a41-c527454c8098.jpg)

OR

API key;
![api _gfitsample](https://cloud.githubusercontent.com/assets/16888650/16792041/82db04ea-48ff-11e6-9068-53dd26c7b97d.jpg)

Thnaks for your help!

Rina

**Comment by rinaha** [comment]
Hi @2dvisio,

We have been waiting for your reply.

Please let us know if you need more information to solve this issue.

Thank you in advance.

Regards,
Rina

**Comment by 2dvisio** [comment]
Hi there.

I have finally found the time to test the plugin once again. The plugin is working fine.
Have you tried starting from scratch and follow all the instructions?

If you have an email address I am more than happy to provide you a test application to try the plugin.

All the best.

**Comment by rinaha** [comment]
Hi @2dvisio,

Thank you for your reply.

It would be great if we could try the plugin with a test application. We've just changed my profile setting to provide you our email address;
https://github.com/rinaha

Looking forwad to hearing from you soon.

Thanks!
Rina

**Comment by rinaha** [comment]
Hello again,

We just wanted to comment that we tried again the plugin by using;
https://github.com/2dvisio/cordova-plugin-googlefit/

Unfortunately we could not get the expected result. We used the last commit to build.

On the other hand, we tried another plugin below which is besed on the plugin above;
https://github.com/dariosalvi78/cordova-plugin-health

We already dealt the HealthKit in our application so we were not planning to use the one cordova-plugin-health, but after the trying we got the expected result (could get the information finally).

What we noticed was;
After the runnning "requestAuthorization" to the information that we want (eg. "steps"), the 
authentication was requested with the user's google account. 

There was no request of the authentication during the testing, but we just wonder if there is the command to send the authentication request in this GoogleFit plugin?

Thank you in advance!
Rina
